### PR TITLE
HPCC-20856 Mark conditional errors as non-throwing

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
This helps with several of the example queries.  (E.g. dschlangen16 - the original query, focusa1, uk18b)

Some examples get larger - e.g., jholt74, but they are generally because more fields are removed, leading to an more classes for representing the meta data.  dabitann is the only example that may possibly deteriorate.

@shamser please review

**Comment by ghalliday** [comment]
The jira contains more details of the logic behind the change.  Please check that as well.

**Review by shamser — APPROVED** [review]
@richardkchapman Code reviewed. Ok.

File: ecl/hql/hqlexpr.cpp [hunk 1]
```diff
@@ -4234,7 +4234,19 @@ void CHqlRealExpression::updateFlagsAfterOperands()
...
-        infoFlags &= ~(HEFnoduplicate|HEFaction|HEFthrowds|HEFthrowscalar|HEFcontainsSkip);
+        //If pure() has extra parameters they allow flags to be selectively removed
+        if (numChildren() > 1)
+        {
+            ForEachChildFrom(i, this, 1)
+            {
+                IHqlExpression * cur = queryChild(i);
+                IAtom * name = cur->queryName();
+                if (name == throwAtom)
+                    infoFlags &= ~(HEFthrowscalar);
```
Comment (richardkchapman) [inline ecl/hql/hqlexpr.cpp]: Should there be an else here (perhaps an error, since the name is not recognized?)

**Comment by richardkchapman** [comment]
@ghalliday One question

File: ecl/hql/hqlexpr.cpp (hunk 1 above)
Comment (ghalliday) [reply ecl/hql/hqlexpr.cpp]: yes, I could throw an exception if unrecognised,

**Comment by ghalliday** [comment]
@richardkchapman else clause added

**Pull request merged**